src/topology_cacher.py

```diff
@@ -0,0 +1,583 @@
+                        for x in resinfo_list
+                        if x.is_ce()
+                    ]
+                    allocation["execute_resource_groups"].append(
+                        {
+                            "ces": ces,
+                            "group_name": group_name,
+                            "local_allocation_id": local_allocation_id,
+                        }
+                    )
+
+                # Done with this allocation
+                allocations.append(allocation)
+
+        # Done with all projects
+        return ret
+
+    #
+    #
+    # Internal
+    #
+    #
+
+    def _get_data(self, endpoint: str) -> ET.Element:
+        """Download XML topology data from from `endpoint` and parse it as an ET.Element.
+        `endpoint` is a path under the topology host, e.g. "/miscproject/xml".
+
+        Returns the parsed data.
+
+        """
+        try:
+            with urlopen(self.topology_base + endpoint) as response:
+                xml_bytes = response.read()  # type: bytes
+        except EnvironmentError as err:
+            raise DataError("Topology query to %s failed" % endpoint) from err
+
+        if not xml_bytes:
+            raise DataError("Topology query to %s returned no data" % endpoint)
+
+        try:
+            element = ET.fromstring(xml_bytes)
+        except (ET.ParseError, UnicodeDecodeError) as err:
+            raise DataError(
+                "Topology query to %s couldn't be parsed" % endpoint
+            ) from err
+
+        return element
+
+
+def safe_element_text(element: Optional[ET.Element]) -> str:
+    return getattr(element, "text", "").strip()
+
+
+def elem2str(element: ET.Element) -> str:
+    return ET.tostring(element, encoding="unicode")
+    # ^^ 'encoding="unicode"' tells ET.tostring() to return an str not a bytes
+
+
+def between(value, minimum, maximum):
+    return max(minimum, min(maximum, value))
+
+
+def main(argv):
+    parser = ArgumentParser(description=__doc__, prog=argv[0])
+    parser.add_argument(
+        "--verbose",
+        "-v",
+        action="count",
+        default=0,
+        help="Say more; can specify twice.",
+    )
+    parser.add_argument(
+        "--quiet",
+        "-q",
+        action="count",
+        default=0,
+        help="Say less; can specify twice.",
+    )
+    parser.add_argument(
+        "--outdir",
+        metavar="DIR",
+        default="/run/topology-cache",
+        help="Directory to write topology files to. [%(default)s]",
+    )
+    parser.add_argument(
+        "--topology",
+        metavar="URL",
+        default=TOPOLOGY,
+        help="Base URL of the Topology service. [%(default)s]",
+    )
+
+    args = parser.parse_args(argv[1:])
+    log.setLevel(
+        between(10 * (args.quiet - args.verbose), logging.DEBUG, logging.CRITICAL)
+    )
+
+    try:
+        os.makedirs(args.outdir, exist_ok=True)
+    except OSError as e:
+        pass  # ¯\_(ツ)_/¯
```

Comment:
if you want to safe characters

```suggestion
        "¯\_(ツ)_/¯"
```

Reply:
I don't, not that much :-)